### ISSUE TITLE
NativeWindow() constructor fail and finalize in GC thread cause crash

### ISSUE THREAD

**Issue by IDXGI** [opentk/opentk#1763]
### Description

Sometime if the hardware is incompatible (such as in virtual machine), the `NativeWindow` construction will throw an exception in `GLFW.CreateWindow` (see the image)
![image](https://github.com/user-attachments/assets/04373234-27ac-4118-8563-6e14f0ce9b3d)
And then, `new NativeWindow()` returns null, we can't get the instance. The instance will be finalize later in GC thread. But `NativeWindow.Dispose()` cannot be called in GC thread, otherwise will throw an exception. This exception will be catched by `AppDomain.UnhandledException`, and result a crash.

Maybe this problem cannot be solved without changing opentk library. 

### Related information 

* Operating System: Windows7 (in virtual machine)
* OpenGL: API 3.3
* OpenTK Version: 5.0.0-pre.13
* Runtime: .NET 9.0

**Comment by NogginBops** [comment]
We are probably not doing to do anything about the potential to get an exception when calling `GLFW.CreateWindow`, but we could handle it more explicitly and have a more intentional exception being thrown.

For the disposing part, we should probably do a null check on `WindowPtr` to check if we even need to destroy the glfw window.
We can fix this for OpenTK 4.9.0.

A potential workaround is to set `GLFWProvider.CheckForMainThread = false`, but that might cause an exception in `GLFW.DestroyWindow`. If a workaround is needed I would try this first.

**Comment by IDXGI** [comment]
Yes, i also think we should fix `Dispose()`. It don't use `CheckForMainThread`, just check directly. So first check `WindowPtr` to ensure initialization is successfully, and then check main thread, then destroy window.

Only check main thread in `Dispose()` when initialization is successful. Because the owner class willl manage it most of time, and will ensure disposing is in the main thread. If init is failed, `NativeWindow.Dispose()` will be called separately by GC.  

This change will affect 4.x and 5.x, and will affect all incompatible situation, so it should be applied to the two major version, i think. 

![image](https://github.com/user-attachments/assets/873f2547-66c6-483c-a652-08c63f4ee2a2)

**Comment by IDXGI** [comment]
OK, thanks. I will test it tomorrow. Also, determining the `ErrorCode` in the constructor doesn't seem to work, when an exception is thrown for a failed creation, it won't be executed below. Here is my StackTrace:

```
2024-11-18 20:50:24.2087 [ERROR] [1] Create context failed. OpenTK.Windowing.GraphicsLibraryFramework.GLFWException: WGL: The driver does not appear to support OpenGL
   at OpenTK.Windowing.Desktop.GLFWProvider.DefaultErrorCallback(ErrorCode errorCode, String description)
   at OpenTK.Windowing.GraphicsLibraryFramework.GLFWNative.glfwCreateWindow(Int32 width, Int32 height, Byte* title, Monitor* monitor, Window* share)
   at OpenTK.Windowing.GraphicsLibraryFramework.GLFW.CreateWindow(Int32 width, Int32 height, String title, Monitor* monitor, Window* share)
   at OpenTK.Windowing.Desktop.NativeWindow..ctor(NativeWindowSettings settings)
   at XDrawing.Internals.Managers.GrGlContextManager.CreateNewContext()
```

**Comment by IDXGI** [comment]
I've finished testing. There is no problem in the `Dispose()` part, everything works fine. Also, you seem to have just turned off the `ErrorCode.PlatformError` exception throw in the `NativeWindow` constructor, but I'm finding that there is still an exception, using `ErrorCode.ApiUnavailable`, is this normal behavior?

![image](https://github.com/user-attachments/assets/1bb2e61b-da89-41e3-9e3b-ebf231c8cb84)

**Comment by NogginBops** [comment]
Ok so I've just merged #1765 which changes the default error message in the default error message callback that says that if the exception is inconvenient you should register your own error message callback. 
This isn't a perfect solution to this issue, but it does have the flexibility where it's easy to fix your own issue and allows you better control over the error callbacks.

We could keep this issue open in case we wanted to do a more fancy default error callbacks that saves an exception and then rethrows it after the glfw function returns, but I'm not sure that is a good solution in reality.

**Comment by IDXGI** [comment]
Okay, thank you for your help :)
